File: decorator/python/example_1/decorator.py
# -*- coding: utf-8 -*-

class Beverage:
    def __init__(self):
        self._description = "Unknown beverage"

    def get_description(self):
        return self._description

    def cost(self):
        pass


class CondimentDecorator(Beverage):

    def get_description(self):
        return None


# Beverages

class Espresso(Beverage):

    def __init__(self):
        self._description = "Espresso"

    def cost(self):
        return 1.99


class HouseBlend(Beverage):

    def __init__(self):
        self._description = "House Blend Coffee"

    def cost(self):
        return .89

class DarkRoast(Beverage):

    def __init__(self):
        self._description = "Dark Roast Coffee"

    def cost(self):
        return 1.99


class Decaf(Beverage):

    def __init__(self):
        self._description = "Decaf Coffee"

    def cost(self):
        return .89

# decorators
class Mocha(CondimentDecorator):

    def __init__(self, beverage_):
        self._beverage = beverage_
        self._description = "Mocha"

    def get_description(self):
        return self._beverage.get_description() + ", " + self._description

    def cost(self):
        return .20 + self._beverage.cost()

class Soy(CondimentDecorator):

    def __init__(self, beverage_):
        self._beverage = beverage_
        self._description = "Soy"

    def get_description(self):
        return self._beverage.get_description() + ", " + self._description

    def cost(self):
        return .35 + self._beverage.cost()


class Whip(CondimentDecorator):

    def __init__(self, beverage_):
        self._beverage = beverage_
        self._description = "Whip"

    def get_description(self):
        return self._beverage.get_description() + ", " + self._description

    def cost(self):
        return .45 + self._beverage.cost()


def main():
    beverage = Espresso()
    print(beverage.get_description() +" $" + str(beverage.cost()))

    beverage2 = DarkRoast()
    beverage2 = Mocha(beverage2)
    beverage2 = Mocha(beverage2)
    beverage2 = Whip(beverage2)
    print(beverage2.get_description() +" $" + str(beverage2.cost()))

    beverage3 = HouseBlend()
    beverage3 = Soy(beverage3)
    beverage3 = Mocha(beverage3)
    beverage3 = Whip(beverage3)
    print(beverage3.get_description() +" $" + str(beverage3.cost()))


if __name__ == "__main__":
    main()

File: decorator/cpp/example_1/decorator.cpp
#include <iostream>
#include <string>

using std::string;
using std::cout;
using std::endl;

class Beverage
{
protected:
  string description = "Unknown Beverage";
public:

  virtual string getDescription() {
    return this->description;
  }

  virtual double cost() = 0;
};

class CondimentDecorator: public Beverage
{
protected:
  Beverage* beverage;

public:
  virtual string getDescription() = 0;

  ~CondimentDecorator() {
    delete beverage;
  }
};

// Beverages

class Espresso: public Beverage
{
public:
  Espresso() {
    this->description = "Espresso";
  }

  virtual double cost() {
    return 1.99;
  }
};

class HouseBlend: public Beverage
{
public:
  HouseBlend() {
    this->description = "House Blend Coffee";
  }

  virtual double cost() {
    return .89;
  }
};


class DarkRoast: public Beverage
{
public:
  DarkRoast() {
    this->description = "Dark Roast Coffee";
  }

  virtual double cost() {
    return .99;
  }
};

class Decaf: public Beverage
{
public:
  Decaf() {
    this->description = "Decaf Coffee";
  }

  virtual double cost() {
    return .77;
  }
};


// decorators

class Mocha: public CondimentDecorator
{

public:
  Mocha(Beverage* beverage_) {
    this->beverage = beverage_;
  }

  virtual string getDescription() {
    return this->beverage->getDescription() + ", Mocha";
  }

  virtual double cost() {
    return .20 + this->beverage->cost();
  }
};

class Soy: public CondimentDecorator
{

public:
  Soy(Beverage* beverage_) {
    this->beverage = beverage_;
  }

  virtual string getDescription() {
    return this->beverage->getDescription() + ", Soy";
  }

  virtual double cost() {
    return .31 + this->beverage->cost();
  }
};

class Whip: public CondimentDecorator
{

public:
  Whip(Beverage* beverage_) {
    this->beverage = beverage_;
  }

  virtual string getDescription() {
    return this->beverage->getDescription() + ", Whip";
  }

  virtual double cost() {
    return .25 + this->beverage->cost();
  }
};

int main() {
  Beverage* beverage = new Espresso();
  cout << beverage->getDescription() << "$ " << beverage->cost() << endl;
  delete beverage;

  Beverage* beverage2 = new DarkRoast();
  beverage2 = new Mocha(beverage2);
  beverage2 = new Mocha(beverage2);
  beverage2 = new Whip(beverage2);
  cout << beverage2->getDescription() << "$ " << beverage2->cost() << endl;
  delete beverage2;

  Beverage* beverage3 = new DarkRoast();
  beverage3 = new Soy(beverage3);
  beverage3 = new Mocha(beverage3);
  beverage3 = new Whip(beverage3);
  cout << beverage3->getDescription() << "$ " << beverage3->cost() << endl;
  delete beverage3;

  return 0;
}
